File: destack-py/destack/store/memory/wiring.py
from destack.language import (
    Entity,
    IsExtensible,
    IsSpatial,
    Materialization,
    Node,
    NodeReference,
    ScalarType,
    Type,
    TypeCardinality,
    Value,
)
from destack.utils.uuid import UUID

from .core import MemoryRow, MemoryTable

NODE_ID_KEY = str(Node.property("id").id)
NODE_PARENT_PTR_KEY = str(Node.property("parent").id)
NODE_SPACE_PTR_ID = str(IsSpatial.property("space").id)
NODE_DEFINITION_PTR_ID = str(IsExtensible.property("definition").id)

ENTITY_SNAPSHOT_PTR_KEY = str(Entity.property("snapshot").id)
ENTITY_MATERIALIZATION_KEY = str(Entity.property("materialization").id)

NODE_REFERENCE_ID_KEY = str(NodeReference.property("id").id)


def pack_node_row(table: MemoryTable, value: Value) -> MemoryRow:
    """Pack a Value into a MemoryRow."""
    value_packed = value.value
    assert value_packed is not None, f"no value for {value!r}"
    id = UUID(value_packed[NODE_ID_KEY])
    ptr = NodeReference(
        type=table.node_type,
        id=UUID(value_packed[NODE_ID_KEY]),
        space_id=UUID(value_packed[NODE_SPACE_PTR_ID][NODE_REFERENCE_ID_KEY])
        if NODE_SPACE_PTR_ID in value_packed
        else None,
        definition_id=UUID(value_packed[NODE_DEFINITION_PTR_ID][NODE_REFERENCE_ID_KEY])
        if NODE_DEFINITION_PTR_ID in value_packed
        else None,
    )
    parent_ptr = value_packed.get(NODE_PARENT_PTR_KEY)
    if parent_ptr is not None:
        parent_ptr = NodeReference.from_value(parent_ptr)
    snapshot_ptr = value_packed.get(ENTITY_SNAPSHOT_PTR_KEY)
    if snapshot_ptr is not None:
        snapshot_ptr = NodeReference.from_value(snapshot_ptr)
    materialization = value_packed.get(ENTITY_MATERIALIZATION_KEY)
    if materialization is not None:
        materialization = Materialization(materialization)
    row = MemoryRow(
        table=table,
        metatype=table.node_type,
        id=id,
        snapshot_id=snapshot_ptr.id if snapshot_ptr is not None else None,
        materialization=materialization,
        ptr=ptr,
        parent_ptr=parent_ptr,
        value=value_packed,
    )
    return row


def unpack_node_row(table: MemoryTable, row: MemoryRow) -> Value:
    """Unpack a MemoryRow to a Value."""
    type_info = Type(
        cardinality=TypeCardinality.SCALAR,
        scalar_type=ScalarType.NODE_VALUE,
        node_type=row.metatype,
    )
    return Value(type=type_info, value=row.value)

File: destack-ts/src/store/memory/wiring.ts
import {
  Entity,
  IsExtensible,
  IsSpatial,
  Materialization,
  Node,
  NodeReference,
  ScalarType,
  Type,
  TypeCardinality,
  Value,
} from "@destack/language";
import { MemoryRow, MemoryTable } from "@destack/store/memory/core";

const NODE_ID_KEY = String(Node.property("id").id);
const NODE_PARENT_PTR_KEY = String(Node.property("parent").id);
const NODE_SPACE_PTR_KEY = String(IsSpatial.property("space").id);
const NODE_DEFINITION_PTR_KEY = String(IsExtensible.property("definition").id);

const ENTITY_SNAPSHOT_PTR_KEY = String(Entity.property("snapshot").id);
const ENTITY_MATERIALIZATION_KEY = String(Entity.property("materialization").id);

const NODE_REFERENCE_ID_KEY = String(NodeReference.property("id").id);

/**
 * Pack a Value into a MemoryRow.
 */
export function packNodeRow(table: MemoryTable, value: Value): MemoryRow {
  const valuePacked = value.value;
  if (!valuePacked) {
    throw new Error(`no value for ${value.repr()}`);
  }

  const id = String(valuePacked[NODE_ID_KEY]);
  const ptr = new NodeReference({
    type: table.nodeType,
    id: String(valuePacked[NODE_ID_KEY]),
    spaceId:
      NODE_SPACE_PTR_KEY in valuePacked
        ? String(valuePacked[NODE_SPACE_PTR_KEY][NODE_REFERENCE_ID_KEY])
        : null,
    definitionId:
      NODE_DEFINITION_PTR_KEY in valuePacked
        ? String(valuePacked[NODE_DEFINITION_PTR_KEY][NODE_REFERENCE_ID_KEY])
        : null,
  });

  let parentPtr: NodeReference | null = null;
  const parentPtrValue = valuePacked[NODE_PARENT_PTR_KEY];
  if (parentPtrValue !== undefined) {
    parentPtr = NodeReference.fromValue(parentPtrValue);
  }

  let snapshotPtr: NodeReference | null = null;
  const snapshotPtrValue = valuePacked[ENTITY_SNAPSHOT_PTR_KEY];
  if (snapshotPtrValue !== undefined) {
    snapshotPtr = NodeReference.fromValue(snapshotPtrValue);
  }

  let materialization: Materialization | null = null;
  const materializationValue = valuePacked[ENTITY_MATERIALIZATION_KEY];
  if (materializationValue !== undefined) {
    materialization = materializationValue as Materialization;
  }

  const row = new MemoryRow(
    table,
    table.nodeType,
    id,
    snapshotPtr?.id || null,
    materialization,
    ptr,
    parentPtr,
    valuePacked,
  );

  return row;
}

/**
 * Unpack a MemoryRow to a Value.
 */
export function unpackNodeRow(table: MemoryTable, row: MemoryRow): Value {
  const typeInfo = new Type({
    cardinality: TypeCardinality.SCALAR,
    scalarType: ScalarType.NODE_VALUE,
    nodeType: row.nodeType,
  });

  return new Value({
    type: typeInfo,
    value: row.value,
  });
}
